File: .legacy-ts/src-ts/writer/obsidian-writer.ts
import { writeFileSync, existsSync } from "fs";
import { join, basename, extname } from "path";
import type { PipelineResult, PipelineConfig, AttributedSegment } from "../types.js";

const DEFAULT_VAULT_PATH = "/home/banar/Desktop/obsidian/Farinter/07-Reuniones";

function formatSeconds(totalSeconds: number): string {
  const h = Math.floor(totalSeconds / 3600);
  const m = Math.floor((totalSeconds % 3600) / 60);
  const s = Math.floor(totalSeconds % 60);
  return [h, m, s].map((v) => String(v).padStart(2, "0")).join(":");
}

function formatDateFromMs(ms: number): string {
  const d = new Date(ms);
  const year = d.getFullYear();
  const month = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${year}-${month}-${day}`;
}

function extractParticipants(transcription: AttributedSegment[]): string[] {
  const seen = new Set<string>();
  for (const seg of transcription) {
    if (seg.speaker) {
      seen.add(seg.speaker);
    }
  }
  return Array.from(seen).sort();
}

function buildFrontmatter(
  fecha: string,
  audioDurationSec: number,
  participants: string[]
): string {
  const participantsYaml =
    participants.length > 0
      ? `[${participants.map((p) => JSON.stringify(p)).join(", ")}]`
      : "[]";

  return [
    "---",
    "tags: [reunion, transcripcion]",
    `fecha: ${fecha}`,
    "estado: completado",
    `duracion: "${formatSeconds(audioDurationSec)}"`,
    `participantes: ${participantsYaml}`,
    "---",
  ].join("\n");
}

function buildTranscriptionSection(transcription: AttributedSegment[]): string {
  const lines = transcription.map((seg) => {
    const time = formatSeconds(seg.start);
    if (seg.speaker) {
      return `**${seg.speaker}** (${time}): ${seg.text}\n`;
    }
    return `(${time}): ${seg.text}\n`;
  });

  return "## Transcripción\n\n" + lines.join("\n");
}

export function writeObsidianNote(
  result: PipelineResult,
  config: PipelineConfig
): string {
  const vaultPath = config.obsidianVaultPath || DEFAULT_VAULT_PATH;

  if (!existsSync(vaultPath)) {
    throw new Error(`E_VAULT_NOT_FOUND: No se encontró la ruta del vault de Obsidian: ${vaultPath}`);
  }

  const fecha = formatDateFromMs(Date.now());
  const baseName = basename(result.inputFile, extname(result.inputFile));
  const fileName = `${fecha} ${baseName}.md`;
  const outputPath = join(vaultPath, fileName);

  const participants = extractParticipants(result.transcription);

  const sections: string[] = [];

  // Frontmatter
  sections.push(buildFrontmatter(fecha, result.audioDurationSec, participants));

  // Title
  sections.push(`\n# Reunión ${fecha} - ${baseName}\n`);

  // Transcription
  if (result.transcription.length > 0) {
    sections.push("\n" + buildTranscriptionSection(result.transcription));
  }

  // Analysis sections (only if analysis exists)
  const analysis = result.analysis;
  if (analysis) {
    // Resumen
    if (analysis.resumen) {
      sections.push("\n## Resumen\n");
      sections.push(analysis.resumen);
    }

    // Requerimientos
    if (analysis.requerimientos.length > 0) {
      sections.push("\n## Requerimientos\n");
      const reqLines = analysis.requerimientos.map(
        (r) => `- [ ] **${r.id}** (${r.prioridad}): ${r.descripcion}`
      );
      sections.push(reqLines.join("\n"));
    }

    // Accionables
    if (analysis.accionables.length > 0) {
      sections.push("\n## Accionables\n");
      const acLines = analysis.accionables.map((a) => {
        const fechaPart = a.fecha ? ` — ${a.fecha}` : "";
        return `- [ ] **${a.responsable}**: ${a.tarea}${fechaPart}`;
      });
      sections.push(acLines.join("\n"));
    }

    // Decisiones
    if (analysis.decisiones.length > 0) {
      sections.push("\n## Decisiones\n");
      const decLines = analysis.decisiones.map((d) => `- ${d}`);
      sections.push(decLines.join("\n"));
    }
  }

  const content = sections.join("\n") + "\n";
  writeFileSync(outputPath, content, "utf-8");

  return outputPath;
}

File: src/video_tranquitor/writers/obsidian_writer.py
"""Escritor de notas en formato Obsidian Markdown."""

from __future__ import annotations

import json
import os
from datetime import datetime
from pathlib import Path

from video_tranquitor.types import AttributedSegment, PipelineConfig, PipelineResult

DEFAULT_VAULT_PATH = "/home/banar/Desktop/obsidian/Farinter/07-Reuniones"


def _format_seconds(total_seconds: float) -> str:
    total = int(total_seconds)
    h = total // 3600
    m = (total % 3600) // 60
    s = total % 60
    return f"{h:02d}:{m:02d}:{s:02d}"


def _format_date_now() -> str:
    """Devuelve la fecha actual como YYYY-MM-DD."""
    d = datetime.now()
    return f"{d.year}-{d.month:02d}-{d.day:02d}"


def _extract_participants(transcription: list[AttributedSegment]) -> list[str]:
    seen: set[str] = set()
    for seg in transcription:
        if seg.speaker:
            seen.add(seg.speaker)
    return sorted(seen)


def _build_frontmatter(
    fecha: str,
    audio_duration_sec: float,
    participants: list[str],
) -> str:
    participants_yaml = (
        "[" + ", ".join(json.dumps(p) for p in participants) + "]"
    )
    lines = [
        "---",
        "tags: [reunion, transcripcion]",
        f"fecha: {fecha}",
        "estado: completado",
        f'duracion: "{_format_seconds(audio_duration_sec)}"',
        f"participantes: {participants_yaml}",
        "---",
    ]
    return "\n".join(lines)


def _build_transcription_section(transcription: list[AttributedSegment]) -> str:
    lines: list[str] = []
    for seg in transcription:
        time = _format_seconds(seg.start)
        if seg.speaker:
            lines.append(f"**{seg.speaker}** ({time}): {seg.text}\n")
        else:
            lines.append(f"({time}): {seg.text}\n")
    return "## Transcripción\n\n" + "\n".join(lines)


def write_obsidian_note(result: PipelineResult, config: PipelineConfig) -> Path:
    """Escribe una nota Markdown en el vault de Obsidian.

    Args:
        result: Resultado completo del pipeline.
        config: Configuración del pipeline (contiene obsidian_vault_path).

    Returns:
        Ruta absoluta del archivo .md generado.

    Raises:
        FileNotFoundError: Si la ruta del vault no existe.
    """
    vault_path = config.obsidian_vault_path or DEFAULT_VAULT_PATH

    if not os.path.exists(vault_path):
        raise FileNotFoundError(
            f"E_VAULT_NOT_FOUND: No se encontró la ruta del vault de Obsidian: {vault_path}"
        )

    fecha = _format_date_now()
    base_name = Path(result.input_file).stem
    file_name = f"{fecha} {base_name}.md"
    output_path = Path(vault_path) / file_name

    participants = _extract_participants(result.transcription)

    sections: list[str] = []

    # Frontmatter
    sections.append(_build_frontmatter(fecha, result.audio_duration_sec, participants))

    # Título
    sections.append(f"\n# Reunión {fecha} - {base_name}\n")

    # Transcripción
    if result.transcription:
        sections.append("\n" + _build_transcription_section(result.transcription))

    # Secciones de análisis (solo si existe)
    analysis = result.analysis
    if analysis:
        if analysis.resumen:
            sections.append("\n## Resumen\n")
            sections.append(analysis.resumen)

        if analysis.requerimientos:
            sections.append("\n## Requerimientos\n")
            req_lines = [
                f"- [ ] **{r.id}** ({r.prioridad}): {r.descripcion}"
                for r in analysis.requerimientos
            ]
            sections.append("\n".join(req_lines))

        if analysis.accionables:
            sections.append("\n## Accionables\n")
            ac_lines: list[str] = []
            for a in analysis.accionables:
                fecha_part = f" — {a.fecha}" if a.fecha else ""
                ac_lines.append(f"- [ ] **{a.responsable}**: {a.tarea}{fecha_part}")
            sections.append("\n".join(ac_lines))

        if analysis.decisiones:
            sections.append("\n## Decisiones\n")
            dec_lines = [f"- {d}" for d in analysis.decisiones]
            sections.append("\n".join(dec_lines))

    content = "\n".join(sections) + "\n"
    output_path.write_text(content, encoding="utf-8")

    return output_path
